File: js/tests/schema-validation.ts
import { readFileSync } from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

import Ajv, { type ErrorObject, type ValidateFunction } from 'ajv';

export interface SchemaViolation {
  message_type: string;
  schema_name: string;
  validation_errors: string[];
  envelope: Record<string, unknown>;
}

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const SCHEMA_DIR = path.resolve(__dirname, '../../shared/schema');

const PAYLOAD_SCHEMA_MAP: Record<string, string> = {
  'system::init': 'system.init.json',
  'system::resync': 'system.resync.json',
  'system::trigger': 'system.trigger.json',
  'system::ping': 'system.ping.json',
  'chat::message': 'chat.message.json',
  'sandbox::continue': 'sandbox.continue.json',
  'sandbox::stop': 'sandbox.stop.json',
  'escalation::reply': 'escalation.reply.json',
};

const ajv = new Ajv({ allErrors: true, strict: false });
ajv.addFormat('date-time', {
  type: 'string',
  validate: (value: string) => typeof value === 'string' && !Number.isNaN(Date.parse(value)),
});

const validatorCache = new Map<string, ValidateFunction>();

function loadValidator(schemaName: string): ValidateFunction {
  const cached = validatorCache.get(schemaName);
  if (cached) {
    return cached;
  }

  const schemaPath = path.join(SCHEMA_DIR, schemaName);
  const schema = JSON.parse(readFileSync(schemaPath, 'utf-8')) as object;
  const validator = ajv.compile(schema);
  validatorCache.set(schemaName, validator);
  return validator;
}

function formatErrors(errors: ErrorObject[] | null | undefined): string[] {
  return (errors ?? []).map((error) => {
    const location = error.instancePath || '/';
    return `${location} ${error.message ?? 'schema validation failed'}`.trim();
  });
}

export function validateOutboundEnvelope(envelope: Record<string, unknown>): SchemaViolation[] {
  const violations: SchemaViolation[] = [];
  const messageType = typeof envelope.type === 'string' ? envelope.type : 'unknown';

  const envelopeValidator = loadValidator('envelope.json');
  if (!envelopeValidator(envelope)) {
    violations.push({
      message_type: messageType,
      schema_name: 'envelope.json',
      validation_errors: formatErrors(envelopeValidator.errors),
      envelope,
    });
  }

  const payloadSchema = PAYLOAD_SCHEMA_MAP[messageType];
  if (!payloadSchema) {
    return violations;
  }

  const payload = envelope.payload;
  const payloadValidator = loadValidator(payloadSchema);
  if (!payloadValidator(payload)) {
    violations.push({
      message_type: messageType,
      schema_name: payloadSchema,
      validation_errors: formatErrors(payloadValidator.errors),
      envelope,
    });
  }

  return violations;
}

File: python/tests/schema_validation.py
from __future__ import annotations

import json
from functools import lru_cache
from pathlib import Path
from typing import Any

from jsonschema import Draft7Validator, FormatChecker


ROOT = Path(__file__).resolve().parents[2]
SCHEMA_DIR = ROOT / "shared" / "schema"

PAYLOAD_SCHEMA_MAP: dict[str, str] = {
    "system::init": "system.init.json",
    "system::resync": "system.resync.json",
    "system::ping": "system.ping.json",
    "chat::message": "chat.message.json",
    "sandbox::continue": "sandbox.continue.json",
    "sandbox::stop": "sandbox.stop.json",
    "escalation::reply": "escalation.reply.json",
}


class SchemaViolation(dict[str, object]):
    pass


@lru_cache(maxsize=None)
def _load_validator(schema_name: str) -> Draft7Validator:
    schema = json.loads((SCHEMA_DIR / schema_name).read_text(encoding="utf-8"))
    return Draft7Validator(schema, format_checker=FormatChecker())


def _format_errors(errors: list[Any]) -> list[str]:
    messages: list[str] = []
    for error in errors:
        location = "/" + "/".join(str(part) for part in error.absolute_path)
        if location == "/":
            location = "/"
        messages.append(f"{location} {error.message}".strip())
    return messages


def validate_outbound_envelope(envelope: dict[str, object]) -> list[SchemaViolation]:
    violations: list[SchemaViolation] = []
    message_type = envelope.get("type")
    message_type_str = message_type if isinstance(message_type, str) else "unknown"

    envelope_validator = _load_validator("envelope.json")
    envelope_errors = sorted(envelope_validator.iter_errors(envelope), key=str)
    if envelope_errors:
        violations.append(SchemaViolation({
            "message_type": message_type_str,
            "schema_name": "envelope.json",
            "validation_errors": _format_errors(envelope_errors),
            "envelope": envelope,
        }))

    payload_schema = PAYLOAD_SCHEMA_MAP.get(message_type_str)
    if payload_schema is None:
        return violations

    payload_validator = _load_validator(payload_schema)
    payload = envelope.get("payload")
    payload_errors = sorted(payload_validator.iter_errors(payload), key=str)
    if payload_errors:
        violations.append(SchemaViolation({
            "message_type": message_type_str,
            "schema_name": payload_schema,
            "validation_errors": _format_errors(payload_errors),
            "envelope": envelope,
        }))

    return violations
